File: leetcode_java/22. Generate Parentheses.java
class Solution {
    List<String> res = new ArrayList<String>();
    public List<String> generateParenthesis(int n) {
        dfs(n, 0, 0, "");
        return res;       
    }
    private void dfs(int n, int i, int j, String tmp) {
        if(i + j == 2 * n) {
            res.add(tmp);
            return;
        }
        if(i < n) dfs(n, i + 1, j, tmp + '(');
        if(j < i) dfs(n, i, j + 1, tmp + ')');
    }
}

File: leetcode_python/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> [str]:
        res = []
        def dfs(i, j, tmp):
            if not (i or j):
                res.append(tmp)
                return
            if i:
                dfs(i - 1, j, tmp + '(')
            if j > i:
                dfs(i, j - 1, tmp + ')')
        dfs(n, n, "")
        return res




   
